server/mvcc/backend/backend.go

```diff
@@ -221,10 +238,55 @@ func (b *backend) ConcurrentReadTx() ReadTx {
 	defer b.readTx.RUnlock()
 	// prevent boltdb read Tx from been rolled back until store read Tx is done. Needs to be called when holding readTx.RLock().
 	b.readTx.txWg.Add(1)
+
 	// TODO: might want to copy the read buffer lazily - create copy when A) end of a write transaction B) end of a batch interval.
+
+	// inspect/update cache recency iff there's no ongoing update to the cache
+	// this falls through if there's no cache update
+	b.txReadBufferCache.mu.Lock()
+
+	curCache := b.txReadBufferCache.buf
+	curCacheVer := b.txReadBufferCache.bufVersion
+	curBufVer := b.readTx.buf.bufVersion
+
+	isEmptyCache := curCache == nil
+	isStaleCache := curCacheVer != curBufVer
+
+	var buf *txReadBuffer
+	switch {
+	case isEmptyCache:
+		// perform safe copy of buffer while holding "b.txReadBufferCache.mu.Lock"
+		// this is only supposed to run once so there won't be much overhead
+		curBuf := b.readTx.buf.unsafeCopy()
+		buf = &curBuf
+	case isStaleCache:
+		// to maximize the concurrency, try unsafe copy of buffer
+		// release the lock while copying buffer -- cache may become stale again and
+		// get overwritten by someone else.
+		// therefore, we need to check the readTx buffer version again
+		b.txReadBufferCache.mu.Unlock()
+		curBuf := b.readTx.buf.unsafeCopy()
+		b.txReadBufferCache.mu.Lock()
+		buf = &curBuf
+	default:
+		// neither empty nor stale cache, just use the current buffer
+		buf = curCache
+	}
+	// txReadBufferCache.bufVersion can be modified when we doing an unsafeCopy()
+	// as a result, curCacheVer could be no longer the same as
+	// txReadBufferCache.bufVersion
+	if isEmptyCache || curCacheVer == b.txReadBufferCache.bufVersion {
+		// continue if the cache is never set or no one has modified the cache
```

Comment:
My question was, is it possible we will evaluate to `else` here?

```
!isEmptyCache && curCacheVer != b.txReadBufferCache.bufVersion
```

If that's the case, we use `buf = curCache` where the `curCache` is stale?

Reply:
> My question was, is it possible this evaluates to `else` here?
> 
> ```
> !isEmptyCache && curCacheVer != b.txReadBufferCache.bufVersion
> ```
> 
> If that's the case, we use `buf = curCache` where the `curCache` is stale?

when we reached the default condition, the lock is never released, which means it is impossible to have curCacheVer != b.txReadBufferCache.bufVersion condition.

Also, in this default condition, we are neither isEmptyCache nor isStaleCache. That means it is not empty cache. So the expression  `!isEmptyCache && curCacheVer != b.txReadBufferCache.bufVersion` is evaluated to `true && false`

Reply:
> the lock is never released, which means it is impossible to have curCacheVer != b.txReadBufferCache.bufVersion condition.

Yeah understood. But what if it goes to `case isStaleCache` where we unlock before copying the buffer? Can this lead to `staleCache && curCacheVer != b.txReadBufferCache.bufVersion`?

Reply:
> > the lock is never released, which means it is impossible to have curCacheVer != b.txReadBufferCache.bufVersion condition.
> 
> Yeah understood. But what if it goes to `case isStaleCache` where we unlock before copying the buffer? Can this lead to `staleCache && curCacheVer != b.txReadBufferCache.bufVersion`?

yeah. assume we have 2 goroutines A & B trying to running this code at almost the same time. A came first but released the lock in the middle. B came later but gained the lock again and went to line 278 earlier than A.

In this scenario, B will finish updating the `b.txReadBufferCache.bufVersion` before A. when A came here, A would realize the `curCacheVer` is no longer the same as `b.txReadBufferCache.bufVersion`

Reply:
> A would realize the curCacheVer is no longer the same as b.txReadBufferCache.bufVersion

Yes, in such case, we do not update `b.txReadBufferCache.buf` with the copied read buffer, while we still pass the value of the copied buffer to the newly created concurrent read transaction (at the end)? Does this mean we will have `concurrentReadTx.baseReadTx.buf != b.txReadBufferCache.buf` (thus stale)?

Reply:
> > A would realize the curCacheVer is no longer the same as b.txReadBufferCache.bufVersion
> 
> Yes, in such case, we do not update `b.txReadBufferCache.buf` with the copied read buffer, while we still pass the value of the copied buffer to the newly created concurrent read transaction (at the end)? Does this mean we will have `concurrentReadTx.baseReadTx.buf != b.txReadBufferCache.buf` (thus stale -- is this safe)?

I would believe in this case, goroutine A & B are just holding different copies of the same readTx.buf. It would be safe since all of the concurrentReadTx are holding their own copy of the data in the original code.

Reply:
Is that because `b.readTx.buf` is already protected read locks?

```go
func (b *backend) ConcurrentReadTx() ReadTx {
	b.readTx.RLock()
	defer b.readTx.RUnlock()
        ...
```

Which guarantees no write can intervene as long as we hold these lock?

Reply:
I would really like to have @jingyih @ptabor take another look, to make sure I don't miss anything :)

Reply:
> Is that because `b.readTx.buf` is already protected read locks?
> 
> ```go
> func (b *backend) ConcurrentReadTx() ReadTx {
> 	b.readTx.RLock()
> 	defer b.readTx.RUnlock()
>         ...
> ```
> 
> Which guarantees no write can intervene as long as we hold these lock?

Regarding write operations 

```go
func (t *batchTxBuffered) Unlock() {
	if t.pending != 0 {
		t.backend.readTx.Lock() // blocks txReadBuffer for writing.
		t.buf.writeback(&t.backend.readTx.buf)
		t.backend.readTx.Unlock()
		if t.pending >= t.backend.batchLimit {
			t.commit(false)
		}
	}
	t.batchTx.Unlock()
}
```

The writeback() is protected by a write lock so no change to readTx.buf can happen when one is reading its contents.

While multiple concurrent read tx can run the code at the same time and we need the lock `b.txReadBufferCache.mu` to protect this.

Reply:
^ @wilsonwang371 exactly this comment, can we explain this inline, around `default:` case?

Reply:
> ^ @wilsonwang371 exactly this comment, can we explain this inline, around `default:` case?

For example,
Goroutine A:
Grpc -> Compact() -> `resp.Header.Revision = s.kv.Rev()` -> `func (rv *readView) Rev() int64` -> `tr := rv.kv.Read(ConcurrentReadTxMode, traceutil.TODO())` -> `func (s *store) Read(mode ReadTxMode, trace *traceutil.Trace) TxnRead ` -> `tx = s.b.ConcurrentReadTx()`

Goroutine B:
applierv3 -> `func (a *applierV3backend) Txn(ctx context.Context, rt *pb.TxnRequest) (*pb.TxnResponse, *traceutil.Trace, error)` -> `txn = mvcc.NewReadOnlyTxnWrite(a.s.KV().Read(mvcc.ConcurrentReadTxMode, trace))` -> `func (s *store) Read(mode ReadTxMode, trace *traceutil.Trace) TxnRead ` -> `tx = s.b.ConcurrentReadTx()`

In this case, if A finished running first, when B continues, B will realize the `curCacheVer == curBufVer` and that means `!isEmptyCache` and `!isStaleCache`. Then B will continue with the buffer that A set. B will not read a value before A set it because this is protected by `b.txReadBufferCache.mu`

Is this the answer you expect?